File: lesson_5_2/app/src/main/java/com/kalashnyk/denys/samplemvvmapp/repository/server/ServerCommunicator.java
package com.kalashnyk.denys.samplemvvmapp.repository.server;

import android.content.Context;

import com.kalashnyk.denys.samplemvvmapp.repository.database.entity.UserEntity;
import com.kalashnyk.denys.samplemvvmapp.repository.database.pojo.UserResponse;

import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.CompletableTransformer;
import io.reactivex.ObservableTransformer;
import io.reactivex.Single;
import io.reactivex.SingleTransformer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

public class ServerCommunicator {
    private static final int DEFAULT_TIMEOUT = 10;
    private static final int DEFAULT_RETRY_ATTEMPTS = 4;
    private ApiService mService;

    public ServerCommunicator(ApiService service) {
        mService = service;
    }

    public Single<Response<UserResponse>> getAllUsers(){
        return mService.getUsers();
    }

    public Single<UserEntity> getUser(int id){
        return mService.getUserById(id);
    }

    private static <T> SingleTransformer<T, T> singleTransformer() {
        return upstream -> upstream
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .timeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .retry(DEFAULT_RETRY_ATTEMPTS);
    }

    private static <T> ObservableTransformer<T, T> observableTransformer() {
        return observable -> observable
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .timeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .retry(DEFAULT_RETRY_ATTEMPTS);
    }

    private static CompletableTransformer completableTransformer() {
        return upstream -> upstream
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .timeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .retry(DEFAULT_RETRY_ATTEMPTS);
    }
}

File: lesson_9/app/src/main/java/com/kalashnyk/denys/kotlinbindingsample/repository/server/ServerCommunicator.kt
package com.kalashnyk.denys.kotlinbindingsample.repository.server

import android.util.Log
import com.kalashnyk.denys.kotlinbindingsample.repository.database.entity.UserEntity
import com.kalashnyk.denys.kotlinbindingsample.repository.database.pojo.UserResponse
import io.reactivex.ObservableTransformer
import io.reactivex.Single
import io.reactivex.SingleTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import retrofit2.Response
import java.util.concurrent.TimeUnit

class ServerCommunicator(private val mService: ApiService) {

    companion object {
        private val DEFAULT_TIMEOUT = 10
        private val DEFAULT_RETRY_ATTEMPTS = 4L
    }

    fun getAllUsers(): Single<Response<UserResponse>> {
        return mService.getUsers()
            .compose(singleTransformer())
            .doOnError { t: Throwable -> Log.d("ServerCommunicator", t.message) }
    }

    fun getUser(id: Int): Single<UserEntity> {
        return mService.getUserById(id).compose(singleTransformer())
    }

    private fun <T> singleTransformer(): SingleTransformer<T, T> = SingleTransformer {
        it.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .timeout(DEFAULT_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .retry(DEFAULT_RETRY_ATTEMPTS)
    }

    private fun <T> observableTransformer(): ObservableTransformer<T, T> = ObservableTransformer {
        it.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .timeout(DEFAULT_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .retry(DEFAULT_RETRY_ATTEMPTS)
    }
}
